rudder-core/src/main/scala/com/normation/rudder/services/policies/JavascriptEngine.scala

```diff
@@ -0,0 +1,708 @@
+      case sha:String if (sha=="sha512" || sha=="sha-512") => aixSha512(s, salt)
+      case _ => // unknown value, fail
+        throw new NoSuchAlgorithmException(s"Evaluating script 'unix(${algo}, ${s}, ${salt})' failed, as algorithm ${algo} is not recognized")
+    }
+  }
+
+  def auto(algo: String, s: String): String = {
+    algo.toLowerCase match {
+      case "md5" => md5(s)
+      case sha:String if (sha=="sha256" || sha=="sha-256") => sha256(s)
+      case sha:String if (sha=="sha512" || sha=="sha-512") => sha512(s)
+      case _ => // unknown value, fail
+        throw new NoSuchAlgorithmException(s"Evaluating script 'auto(${algo}, ${s})' failed, as algorithm ${algo} is not recognized")
+    }
+  }
+
+  def auto(algo: String, s: String, salt: String): String = {
+    algo.toLowerCase match {
+      case "md5" => aixMd5(s)
+      case sha:String if (sha=="sha256" || sha=="sha-256") => sha256(s, salt)
+      case sha:String if (sha=="sha512" || sha=="sha-512") => sha512(s, salt)
+      case _ => // unknown value, fail
+        throw new NoSuchAlgorithmException(s"Evaluating script 'auto(${algo}, ${s}, ${salt})' failed, as algorithm ${algo} is not recognized")
+    }
+  }
+}
+
+/*
+ * This class provides the Rudder JS lib.
+ * All method signatures are available in JS.
+ *
+ * This lib is intended to be bound to the "rudder" namespace,
+ * so that one can access:
+ * ## Under the "rudder.hash" namespace, simple hashing methods:
+ * - rudder.hash.md5(value)
+ * - rudder.hash.sha256(value)
+ * - rudder.hash.sha512(value)
+ *
+ * ## Under the "rudder.password" namespace, salted hashing functions
+ *    compatible with Unix crypt (Linux, BSD...) or AIX
+ * ### Automatically chosen based on the node type:
+ * - rudder.password.md5(value [, salt])
+ * - rudder.password.sha256(value [, salt])
+ * - rudder.password.sha512(value [, salt])
+ *
+ * ### Generates Unix crypt password compatible hashes:
+ * - rudder.password.cryptMd5(value [, salt])
+ * - rudder.password.cryptSha256(value [, salt])
+ * - rudder.password.cryptSha512(value [, salt])
+ *
+ * ### Generates AIX password compatible hashes:
+ * - rudder.password.aixMd5(value [, salt])
+ * - rudder.password.aixSha256(value [, salt])
+ * - rudder.password.aixSha512(value [, salt])
+ *
+ * ### Public methods (advertised to the users, fallback to the previous methods)
+ * - rudder.password.auto(algo, password [, salt])
+ * - rudder.password.unix(algo, password [, salt])
+ * - rudder.password.aix(algo, password [, salt])
+ *
+ *   where algo can be MD5, SHA-512, SHA-256 (case insensitive, with or without -)
+ *   * auto automatically choose the encryption based on the node type
+ *   * unix generated Unix crypt password compatible hashes (Linux, BSD, ...)
+ *   * aix generates AIX password compatible hashes
+ */
+final class JsRudderLibImpl(
+  hashKind: HashOsType
+) {
+
+  ///// simple hash algorithms /////
+  private val hash = new JsLibHash()
```

Comment:
i suspect this is for visibility issue. With the getHash, it gives access to hash - but if hash is public, there might be some java magic  going on